### ISSUE TITLE
型アノテーションの補完

### ISSUE THREAD

**Issue by kurorosu** [kurorosu/pochitrain#350]
## Summary

一部のメソッドで型アノテーションが不完全 (`dict`, `list` のみで値型が未指定).

## Background

- mypy strict モードでは検出されるが, テスト用の relaxed モードでは見逃される
- IDE の補完精度が低下する

## Current Implementation

| ファイル | メソッド | 現状 | 修正後 |
|---------|---------|------|--------|
| `pochi_dataset.py` | `get_class_counts()` | `-> dict` | `-> dict[str, int]` |
| `utils/directory_manager.py` | `save_dataset_paths()` | `train_paths: list` | `train_paths: list[str]` |
| `utils/directory_manager.py` | `save_dataset_paths()` | `val_paths: Optional[list]` | `val_paths: Optional[list[str]]` |
| `training/layer_wise_lr/param_group_builder.py` | `build()` 内 | `layer_params: dict[str, list]` | `layer_params: dict[str, list[torch.nn.Parameter]]` |

## Proposed Changes

上記4箇所の型アノテーションを具体的な値型に補完する.

## Acceptance Criteria

- 型アノテーションが具体的な値型を含んでいること
- `uv run mypy pochitrain` がパスすること
- 既存テストがすべてパスすること

## Labels

`refactor`

## Branch

`refactor/type-annotations`
